core/src/main/scala/cats/instances/vector.scala

```diff
@@ -138,38 +141,42 @@ trait VectorInstances extends cats.kernel.instances.VectorInstances {
        * This avoids making a very deep stack by building a tree instead
        */
       override def traverse_[G[_], A, B](fa: Vector[A])(f: A => G[B])(implicit G: Applicative[G]): G[Unit] = {
-        // the cost of this is O(size)
-        // c(n) = 1 + 2 * c(n/2)
-        // invariant: size >= 1
-        def runHalf(size: Int, idx: Int): Eval[G[Unit]] =
-          if (size > 1) {
-            val leftSize = size / 2
-            val rightSize = size - leftSize
-            runHalf(leftSize, idx)
-              .flatMap { left =>
-                val right = runHalf(rightSize, idx + leftSize)
-                G.map2Eval(left, right) { (_, _) => () }
+        G match {
+          case x: StackSafeMonad[G] => Traverse.traverse_Directly(fa)(f)(x)
+          case _                    =>
+            // the cost of this is O(size)
+            // c(n) = 1 + 2 * c(n/2)
+            // invariant: size >= 1
+            def runHalf(size: Int, idx: Int): Eval[G[Unit]] =
+              if (size > 1) {
+                val leftSize = size / 2
+                val rightSize = size - leftSize
+                runHalf(leftSize, idx)
+                  .flatMap { left =>
+                    val right = runHalf(rightSize, idx + leftSize)
```

Comment:
same comment about the possibility to short circuit the right with `Eval.defer(runHalf(rightSize, idx + leftSize))` but that would add cost to the cases where we do need to evaluate everything, which is probably the most common case.